src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1735,6 +1717,62 @@ void MarkField (FieldDefinition field, in DependencyInfo reason, in MessageOrigi
 			}
 		}
 
+		bool ShouldWarnForReflectionAccessToCompilerGeneratedCode (FieldDefinition field)
+		{
+			if (!CompilerGeneratedState.IsNestedFunctionOrStateMachineMember (field))
+				return false;
+
+			// No need to warn if it's already covered by the Requires attribute or explicit annotations on the field.
+			if (Annotations.DoesFieldRequireUnreferencedCode (field, out _) || Annotations.FlowAnnotations.ShouldWarnWhenAccessedForReflection (field))
```

Comment:
Maybe we should restructure the code below - basically do the "Reflection access on compiler generated code" last (currently we do it "first"). And then remember if we already produced a warning due to RUC or annotation for the field and then simply avoid the compiler generated check completely.

The current implementation requires the two methods to be in sync, which is not robust. And it also performs the same check multiple times.